bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/discovery/internal/BluetoothAddressLocker.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bluetooth.discovery.internal;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.openhab.binding.bluetooth.BluetoothAddress;
+
+/**
+ * The {@link BluetoothAddressLocker} handles global locking of BluetoothAddress.
+ * This is used to make sure that devices with handlers are not connected to during discovery.
+ *
+ * @author Connor Petty - Initial Contribution
+ */
+public class BluetoothAddressLocker {
+
+    private static Map<BluetoothAddress, LockReference> locks = new ConcurrentHashMap<>();
+
+    public static void lock(BluetoothAddress address) {
+        locks.compute(address, (addr, oldRef) -> {
+            LockReference ref = oldRef;
+            if (ref == null) {
+                ref = new LockReference();
+            }
+            ref.count++;
```

Comment:
Is this only for keeping track of the number of locks? Wouldn't `.getHoldCount()` also work? And why are there several  locks for the same object?

Reply:
The counters are meant to keep track of outside references to the lock, so that once there are no references, the lock can be removed from the map and then garbage collected. Since bluetooth address are dynamic and constantly changing, it is necessary to design it like this for long term stability of the runtime.
The entire reason I created this was to prevent connection discovery from connecting to a device that has a handler. Since a handler should have de facto control over the device, discovery shouldn't change its state.
This locking works in conjunction with `BluetoothDevice.hasListeners()` to determine whether or not a device is in use.
Although, after taking the time to write this all out has made me realize that there is a flaw in using "hasListeners" for this purpose, so I'll have to rewrite that logic.

Reply:
Is this blocking the merge or can it be added alter without breaking something until then

Reply:
Not blocking at all, just finished adding it. Would have done so yesterday but was sick 🤷‍♂